File: engine.py
import random

from cards import DECK, beats, rank, suit


class Durak:
    def __init__(self):
        deck = list(DECK)
        random.shuffle(deck)

        self.hands = [sorted(deck[0:6]), sorted(deck[6:12])]

        self.talon = deck[12:]
        self.trump_card = self.talon[0]
        self.trump = suit(self.trump_card)

        self.attacker = self.first_attacker()
        self.table = []
        self.discard = []
        self.phase = "ATTACK"
        self.max_attacks = min(6, len(self.hands[1 - self.attacker]))

        self.durak = None

    def copy(self):
        game = Durak.__new__(Durak)
        game.hands = [self.hands[0][:], self.hands[1][:]]
        game.talon = self.talon[:]
        game.trump = self.trump
        game.trump_card = self.trump_card
        game.table = [row[:] for row in self.table]
        game.discard = self.discard[:]
        game.attacker = self.attacker
        game.phase = self.phase
        game.max_attacks = self.max_attacks
        game.durak = self.durak
        return game

    def first_attacker(self):
        best_player = 0
        best_trump = None

        for player in (0, 1):
            hand = self.hands[player]
            for card in hand:
                if suit(card) == self.trump and (best_trump is None or card < best_trump):
                    best_trump = card
                    best_player = player

        return best_player

    def get_actions(self):
        if self.phase == "OVER":
            return []

        if self.phase == "DEFEND":
            defender = 1 - self.attacker
            attack_card = self.table[-1][0]
            actions = [("defend", card) for card in self.hands[defender] if beats(attack_card, card, self.trump)]
            actions.append(("take", None))
            return actions
        
        if self.phase == "ATTACK" and not self.table:
            return [("attack", card) for card in self.hands[self.attacker]]
        
        actions = []
        if len(self.table) < self.max_attacks:
            table_ranks = set()
            for row in self.table:
                for card in row:
                    if card is not None:
                        table_ranks.add(rank(card))
            actions = [("attack", card) for card in self.hands[self.attacker] if rank(card) in table_ranks]

        actions.append(("end", None))
        return actions

    def apply(self, action):
        action_type, card = action

        if action_type == "attack":
            self.hands[self.attacker].remove(card)
            self.table.append([card, None])
            if self.phase != "TAKING":
                self.phase = "DEFEND"

        elif action_type == "defend":
            self.hands[1 - self.attacker].remove(card)
            self.table[-1][-1] = card
            self.phase = "ATTACK"

        elif action_type == "take":
            self.phase = "TAKING"

        elif action_type == "end":
            cards = []
            for row in self.table:
                for card in row:
                    if card is not None:
                        cards.append(card)

            defender = 1 - self.attacker
            if self.phase == "TAKING":
                self.hands[defender].extend(cards)
            else:
                self.discard.extend(cards)

            self.table.clear()
            self.draw(self.attacker)
            self.draw(defender)

            if not self.talon:
                if not self.hands[self.attacker] or not self.hands[defender]:
                    self.phase = "OVER"
                    if not self.hands[self.attacker] and not self.hands[defender]:
                        self.durak = None
                    elif not self.hands[self.attacker]:
                        self.durak = defender
                    else:
                        self.durak = self.attacker
                    return

            if self.phase != "TAKING":
                self.attacker = defender

            self.phase = "ATTACK"
            self.max_attacks = min(6, len(self.hands[1 - self.attacker]))

    def draw(self, player):
        hand = self.hands[player]
        while len(hand) < 6 and self.talon:
            hand.append(self.talon.pop())

    def current_player(self):
        if self.phase == "DEFEND":
            return 1 - self.attacker
        return self.attacker

    def hidden_from(self, player):
        seen = set(self.hands[player])
        seen.update(self.discard)
        for row in self.table:
            for card in row:
                if card is not None:
                    seen.add(card)
        return [card for card in DECK if card not in seen]

File: rust/src/engine.rs
use crate::cards::{Card, DECK_SIZE, beats, suit};
use crate::hand::Hand;
use rand::rngs::SmallRng;
use rand::seq::SliceRandom;

pub const HAND_SIZE: usize = 6;

#[derive(Clone, Copy, PartialEq, Eq, Debug)]
pub enum Action {
    Attack(Card),
    Defend(Card),
    Take,
    End,
}

#[derive(Clone, Copy, PartialEq, Eq, Debug)]
pub enum Phase {
    Attack,
    Defend,
    Taking,
    Over,
}

#[derive(Clone)]
pub struct Durak {
    pub hands: [Hand; 2],
    pub talon: Vec<Card>,
    pub trump_card: Card,
    pub trump: u8,
    pub attacker: usize,
    pub table: Vec<(Card, Option<Card>)>,
    pub discard: Hand,
    pub phase: Phase,
    pub max_attacks: usize,
    pub durak: Option<usize>,
}

impl Durak {
    pub fn new(rng: &mut SmallRng) -> Durak {
        let mut deck: Vec<Card> = (0..DECK_SIZE as Card).collect();
        deck.shuffle(rng);

        let mut first_hand = deck[0..HAND_SIZE].to_vec();
        let mut second_hand = deck[HAND_SIZE..2 * HAND_SIZE].to_vec();
        first_hand.sort();
        second_hand.sort();

        let hands = [Hand::from_cards(first_hand), Hand::from_cards(second_hand)];
        let talon = deck[2 * HAND_SIZE..].to_vec();
        let trump_card = talon[0];
        let trump = suit(trump_card);
        let attacker = first_attacker(&hands, trump);
        let max_attacks = HAND_SIZE.min(hands[1 - attacker].len());

        Durak {
            hands,
            talon,
            trump_card,
            trump,
            attacker,
            table: Vec::new(),
            discard: Hand::new(),
            phase: Phase::Attack,
            max_attacks,
            durak: None,
        }
    }

    pub fn get_actions(&self) -> Vec<Action> {
        if self.phase == Phase::Over {
            return Vec::new();
        }

        if self.phase == Phase::Defend {
            let defender = 1 - self.attacker;
            let attack_card = self.table[self.table.len() - 1].0;
            let mut actions = Vec::new();

            for card in self.hands[defender].cards() {
                if beats(attack_card, card, self.trump) {
                    actions.push(Action::Defend(card));
                }
            }
            actions.push(Action::Take);
            return actions;
        }

        if self.phase == Phase::Attack && self.table.is_empty() {
            let mut actions = Vec::new();
            for card in self.hands[self.attacker].cards() {
                actions.push(Action::Attack(card));
            }
            return actions;
        }

        Vec::new()
    }
}

fn first_attacker(hands: &[Hand; 2], trump: u8) -> usize {
    let mut best_player = 0;
    let mut best_trump = None;

    for player in 0..2 {
        for card in hands[player].cards() {
            if suit(card) == trump && (best_trump.is_none() || card < best_trump.unwrap()) {
                best_trump = Some(card);
                best_player = player;
            }
        }
    }

    best_player
}
